indutor: fix issue of compute index_expr range

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103147

For the CPU inductor side, there has an optimization to convert ```int64``` index_expr to ```int32``` for good performance(https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/cpp.py#L2034), but for ```ModularIndexing``` exp, we replace it as division(https://github.com/pytorch/pytorch/blob/main/torch/_inductor/optimize_indexing.py#L73, ```ModularIndexing``` doesn't have derivative) to compute derivative and then compute the expr's value range, there may meet issue which the min value may greater than the max value(```ModularIndexing(513*i2 + i3 + 262400, 512, 513), with vars_ranges is {i2: ValueRanges(lower=0, upper=256), i3: ValueRanges(lower=0, upper=513)}```).

One solution is that we don't replace ```ModularIndexing```, but it can't get the value range.
Another solution is that return ```inf``` range when the min val is great than the max val.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78

XiaobingSuper committed:
indutor: fix issue of compute index_expr range

[ghstack-poisoned]

XiaobingSuper commented:
I just meet one index expr which only has one ```ModularIndexing```, for a complex expr has ```ModularIndexing``` and other operatoions,  I don't think it is ok using the current method to get the value range.

jgong5 commented on torch/_inductor/optimize_indexing.py:
```diff
@@ -125,7 +125,13 @@ def indexing_div_rep(x, y):
             },
         )
-        return ValueRanges(min_val, max_val)
+        # min_value may be greater than max_value, such as ModularIndexing(513*i2 + i3 + 262400, 512, 513),
+        # with vars_ranges is {i2: ValueRanges(lower=0, upper=256), i3: ValueRanges(lower=0, upper=513)}.
```
I guess a better approach is to deduce the range with the divisor. For example, we may create a new symbol with the range [0, divisor-1] suppose the divisor is constant, and put this range to the vars_ranges to calculate the min/max value with the algorithm in this function.

XiaobingSuper committed:
Update on "indutor: fix issue of compute index_expr range"


For the CPU inductor side, there has an optimization to convert ```int64``` index_expr to ```int32``` for good performance(https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/cpp.py#L2034), but for ```ModularIndexing``` exp, we replace it as division(https://github.com/pytorch/pytorch/blob/main/torch/_inductor/optimize_indexing.py#L73, ```ModularIndexing``` doesn't have derivative) to compute derivative and then compute the expr's value range, there may meet issue which the min value may greater than the max value(```ModularIndexing(513*i2 + i3 + 262400, 512, 513), with vars_ranges is {i2: ValueRanges(lower=0, upper=256), i3: ValueRanges(lower=0, upper=513)}```).

One solution is that we don't replace ```ModularIndexing```, but it can't get the value range.
Another solution is that return ```inf``` range when the min val is great than the max val.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

jgong5 commented on torch/_inductor/optimize_indexing.py:
```diff
@@ -64,15 +65,22 @@ def range_expressable_in_32_bits(range):
+    if len(symbols) == 0:
         return ValueRanges(expr, expr)
 
+    vars_ranges_temp = vars_ranges.copy()
```
simpler to do `vars_ranges = vars_ranges.copy()`?

ngimel commented:
We should just return (0, z-1) range for ModularIndexing, and not go through derivative computation. cc @eellison

XiaobingSuper commented:
> We should just return (0, z-1) range for ModularIndexing, and not go through derivative computation. cc @eellison

Yes, for a simple expr which only has ModularIndexing, it is ok just to return (0, z-1), but for a complex expr, I think we need still to use derivative and replace ModularIndexing as a variable that has a range (0, z-1). @eellison

eellison commented on torch/_inductor/optimize_indexing.py:
```diff
@@ -64,15 +65,22 @@ def range_expressable_in_32_bits(range):
         def mod_indexing_rep(x, y, z):
             if z.is_constant():
-                return x / y
+                new_var = sympy_symbol("mod_index" + f"{next(cnt)}")
```
Should we check if `x / y` has a range <= `z` and return `x / y` in that case ?

XiaobingSuper replied:
I we want to return ```x/y```,  it needs to check ```z``` is positive or not, and ```x/y``` ranges are the same sign, for example, if ```x/y``` ranges are ```[-2, 2]```, ```z ``` is ```4```, we can't direct return ```x/y```. if we want to return ```x/y```, we may need to add consider many conditions, I think using z's range is ok even if the range is huge for some cases.

XiaobingSuper replied:
Yes, added.

eellison reviewed (commented):
possible to add test ?

ngimel commented:
cc @lezcano, does this overlap with #102722?

lezcano commented:
Yeah, https://github.com/pytorch/pytorch/pull/102722 is strictly better than this, but this is better than the current state so I'm fine with merging this one.

FWIW, I'm going to be on PTO for the next 3 weeks, so let's merge this one now and I'll have the other one ready when I'm back.

XiaobingSuper committed:
Update on "indutor: fix issue of compute index_expr range"


For the CPU inductor side, there has an optimization to convert ```int64``` index_expr to ```int32``` for good performance(https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/cpp.py#L2034), but for ```ModularIndexing``` exp, we replace it as division(https://github.com/pytorch/pytorch/blob/main/torch/_inductor/optimize_indexing.py#L73, ```ModularIndexing``` doesn't have derivative) to compute derivative and then compute the expr's value range, there may meet issue which the min value may greater than the max value(```ModularIndexing(513*i2 + i3 + 262400, 512, 513), with vars_ranges is {i2: ValueRanges(lower=0, upper=256), i3: ValueRanges(lower=0, upper=513)}```).

One solution is that we don't replace ```ModularIndexing```, but it can't get the value range.
Another solution is that return ```inf``` range when the min val is great than the max val.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

XiaobingSuper commented:
> possible to add test ?

I have tried to get a smaller repro test case but failed. Or do I need to test ```ModularIndexing(513*i2 + i3 + 262400, 512, 513), with vars_ranges is {i2: ValueRanges(lower=0, upper=256), i3: ValueRanges(lower=0, upper=513)}``` using ```get_expr_range```.

XiaobingSuper committed:
Update on "indutor: fix issue of compute index_expr range"


For the CPU inductor side, there has an optimization to convert ```int64``` index_expr to ```int32``` for good performance(https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/cpp.py#L2034), but for ```ModularIndexing``` exp, we replace it as division(https://github.com/pytorch/pytorch/blob/main/torch/_inductor/optimize_indexing.py#L73, ```ModularIndexing``` doesn't have derivative) to compute derivative and then compute the expr's value range, there may meet issue which the min value may greater than the max value(```ModularIndexing(513*i2 + i3 + 262400, 512, 513), with vars_ranges is {i2: ValueRanges(lower=0, upper=256), i3: ValueRanges(lower=0, upper=513)}```).

One solution is that we don't replace ```ModularIndexing```, but it can't get the value range.
Another solution is that return ```inf``` range when the min val is great than the max val.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot rebase

XiaobingSuper committed:
Update on "indutor: fix issue of compute index_expr range"


For the CPU inductor side, there has an optimization to convert ```int64``` index_expr to ```int32``` for good performance(https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/cpp.py#L2034), but for ```ModularIndexing``` exp, we replace it as division(https://github.com/pytorch/pytorch/blob/main/torch/_inductor/optimize_indexing.py#L73, ```ModularIndexing``` doesn't have derivative) to compute derivative and then compute the expr's value range, there may meet issue which the min value may greater than the max value(```ModularIndexing(513*i2 + i3 + 262400, 512, 513), with vars_ranges is {i2: ValueRanges(lower=0, upper=256), i3: ValueRanges(lower=0, upper=513)}```).

One solution is that we don't replace ```ModularIndexing```, but it can't get the value range.
Another solution is that return ```inf``` range when the min val is great than the max val.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

eellison commented on test/inductor/test_cpu_repro.py:
```diff
@@ -387,6 +387,26 @@ def fn(x):
             (torch.randn(8),),
         )
 
+    @patch("torch.cuda.is_available", lambda: False)
```
why do you need this ?

XiaobingSuper replied:
Remove it.

XiaobingSuper committed:
Update on "indutor: fix issue of compute index_expr range"


For the CPU inductor side, there has an optimization to convert ```int64``` index_expr to ```int32``` for good performance(https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/cpp.py#L2034), but for ```ModularIndexing``` exp, we replace it as division(https://github.com/pytorch/pytorch/blob/main/torch/_inductor/optimize_indexing.py#L73, ```ModularIndexing``` doesn't have derivative) to compute derivative and then compute the expr's value range, there may meet issue which the min value may greater than the max value(```ModularIndexing(513*i2 + i3 + 262400, 512, 513), with vars_ranges is {i2: ValueRanges(lower=0, upper=256), i3: ValueRanges(lower=0, upper=513)}```).

One solution is that we don't replace ```ModularIndexing```, but it can't get the value range.
Another solution is that return ```inf``` range when the min val is great than the max val.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper committed:
Update on "indutor: fix issue of compute index_expr range"


For the CPU inductor side, there has an optimization to convert ```int64``` index_expr to ```int32``` for good performance(https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/cpp.py#L2034), but for ```ModularIndexing``` exp, we replace it as division(https://github.com/pytorch/pytorch/blob/main/torch/_inductor/optimize_indexing.py#L73, ```ModularIndexing``` doesn't have derivative) to compute derivative and then compute the expr's value range, there may meet issue which the min value may greater than the max value(```ModularIndexing(513*i2 + i3 + 262400, 512, 513), with vars_ranges is {i2: ValueRanges(lower=0, upper=256), i3: ValueRanges(lower=0, upper=513)}```).

One solution is that we don't replace ```ModularIndexing```, but it can't get the value range.
Another solution is that return ```inf``` range when the min val is great than the max val.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge
